### PR TITLE
Connection protocol

### LINKED ISSUE #2: Implement fixed time polling loop for network communication

### PR DESCRIPTION
**Pull request by Henkecool**
# Connection protocol

The base connection protocol structure and the protocol estabished so far is is implemented.

* Add RemoteData abstraction
* Add WIP Car model
* Add WIP Server class for handeling packets
* Add listeners for Acknowledgement, Errors, Exceptions and Updates
* Add runnable fixed time puller
* Add CommunicationPacket class for representing data packets
* Add PacketList class for containting list of CommunicationPackets
* Add RequestBuilder class for convenient building of PacketLists to be sent by server
* Add enumerations for a bunch of things
* Add RequestIDBroker class for generating and keeping track of used packet ID:s
* Add two test mockups

## Relevant issues
Fixes #2. #3 now WIP.

## Testing
Basic testing with mockup server was conducted.

## Assignees
Tag and assign the collaborators responsible for solving the pull request and the collaborator that is responsible for giving a code review.

### CODE REVIEW

File: src/enums/ControlMode.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    MANUAL,
+    SEMI_AUTO,
+    FULL_AUTO;
+
+    public byte code() {
+        return (byte) this.ordinal();
+    }
+
+    public static ControlMode fromByte(byte b) {
+        return ControlMode.values()[b];
```
Comment (angloth) [inline src/enums/ControlMode.java:15]: Might need check of the input byte to avoid index error

File: src/enums/PacketCommand.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            ord = (b & 0xff);
+        }
+        else if ((b & 0xff) < ERROR_BASE) {
+            ord = (b & 0xff) - CENTRAL_RESPONS_BASE + NUMBER_OF_REMOTE_REQUESTS + 1;
+        }
+        else {
+            ord = (b & 0xff) - ERROR_BASE + NUMBER_OF_REMOTE_REQUESTS + NUMBER_OF_CENTRAL_RESPONSES + 1;
+        }
+
+        return PacketCommand.values()[ord];
```
Comment (angloth) [inline src/enums/PacketCommand.java:156]: Might need an index check on 'ord' aswell?

File: src/remote/Car.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    public RemoteData<ControlMode> controlMode;
+
+    public RemoteData<Float> temperature;
+
+    // TODO change to float
+    public RemoteData<Integer> distanceToLeft;
+    public RemoteData<Integer> distanceToMiddle;
+    public RemoteData<Integer> distanceToRight;
+
+    public Car() {
```
Comment (angloth) [inline src/remote/Car.java]: Constructor should be private for a true singleton.

File: src/remote/RequestBuilder.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        //addDatalessRequest(PacketCommand.REQUEST_STATUS);
+    }
+
+//    public void addSendMapRequest(Map map) {
+//        addRequest(Enums.PacketType.SEND_MAP, map.toBytes());
+//    }
+
+//    public void addSendParametersRequest(ControlParameters param) {
+//        addRequest(Enums.PacketType.SEND_PARAMETERS, param.toBytes());
+//    }
```
Comment (angloth) [inline src/remote/RequestBuilder.java:46]: Would prefer to remove outcommented code.

File: src/remote/RequestBuilder.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    public void addTurnRequest(int turn) {
+        if(turn >= 0x7f || turn <= -0x7f)
+            throw new IllegalArgumentException("Given turn is out of range (" + turn + ")");
+
+        addRequest(PacketCommand.REQUEST_TURN, new byte[]{(byte) turn});
+    }
+
+//    public void addSendRoute(Route route) {
+//        addRequest(Enums.PacketType.SEND_NEW_ROUTE, route.toBytes());
+//    }
```
Comment (angloth) [inline src/remote/RequestBuilder.java:69]: Outcommented code

File: src/remote/RequestIDBroker.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    public int getID() {
+        while(true) {
+            int prop = ID_RANGE_MIN + random.nextInt(ID_RANGE_MAX);
+
+            if (!usedIDs.contains(prop)) {
+                usedIDs.add(prop);
+                return prop;
+            }
+        }
+    }
```
Comment (angloth) [inline src/remote/RequestIDBroker.java:35]: Lol

File: src/remote/Server.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    private RequestIDBroker requestIDBroker;
+    private RequestBuilder requestBuilder;
+
+    private List<AcknowledgementListener> acknowledgementListeners;
+    private List<ErrorListener> errorListeners;
+    private List<ExceptionListener> exceptionListeners;
+
+    private volatile boolean builderLocked = true;
+
+    public Server() {
```
Comment (angloth) [inline src/remote/Server.java]: Public constructor for Singleton

**Review by angloth — CHANGES_REQUESTED** [review]
Ok boomer

File: src/remote/RequestBuilder.java (hunk 4 above)
Comment (Henkecool) [reply src/remote/RequestBuilder.java:46]: This is WIP. Will be resolved when communication protocol is fully established

File: src/remote/RequestBuilder.java (hunk 5 above)
Comment (Henkecool) [reply src/remote/RequestBuilder.java:69]: This is WIP. Will be resolved when communication protocol is fully established

File: src/remote/RequestIDBroker.java (hunk 6 above)
Comment (Henkecool) [reply src/remote/RequestIDBroker.java:35]: xD

**Review by Henkecool — COMMENTED** [review]
New changes added to resolve pull request comments.

**Review by Bence657 — APPROVED** [review]
Ok






























Boomer

**Review by angloth — APPROVED** [review]
OK

**Pull request merged**